src/java.desktop/unix/classes/sun/awt/screencast/TokenStorage.java

```diff
@@ -0,0 +1,411 @@
+        // with exact matches at the beginning.
+        LinkedHashSet<TokenItem> result = new LinkedHashSet<>();
+
+        Set<Map.Entry<Object, Object>> entries;
+        synchronized (PROPS_LOCK) {
+            entries = PROPS.entrySet();
+        }
+
+        Set<String> malformed = new LinkedHashSet<>();
+
+        List<TokenItem> allTokenItems = entries
+                .stream()
+                .map(o -> {
+                    String token = String.valueOf(o.getKey());
+                    TokenItem tokenItem =
+                            TokenItem.parse(token, o.getValue());
+                    if (tokenItem == null) {
+                        malformed.add(token);
+                    }
+                    return tokenItem;
+                })
+                .filter(obj -> !Objects.isNull(obj))
+                .toList();
+
+        removeMalformedRecords(malformed);
+
+        // 1. Try to find exact matches
+        for (TokenItem tokenItem : allTokenItems) {
+            if (tokenItem != null
+                    && tokenItem
+                    .hasAllScreensWithExactMatch(affectedScreenBounds)) {
+                result.add(tokenItem);
+            }
+        }
+
+        if (SCREENCAST_DEBUG) {
+            System.out.println("// getTokens exact matches 1. " + result);
+        }
+
+        // 2. Try screens of the same size but in different locations,
+        // screens may have been moved while the token is still valid
+        List<Dimension> dimensions = affectedScreenBounds
+                .stream()
+                .map(rectangle -> new Dimension(
+                        rectangle.width,
+                        rectangle.height
+                ))
+                .collect(Collectors.toCollection(ArrayList::new));
+
+        for (TokenItem tokenItem : allTokenItems) {
+            if (tokenItem != null
+                    && tokenItem.hasAllScreensOfSameSize(dimensions)) {
+                result.add(tokenItem);
+            }
+        }
+
+        if (SCREENCAST_DEBUG) {
+            System.out.println("// getTokens same sizes 2. " + result);
+        }
+
+        return result;
+    }
+
+    private static void removeMalformedRecords(Set<String> malformedRecords) {
+        if (!isWritable()
+                || malformedRecords == null
+                || malformedRecords.isEmpty()) {
+            return;
+        }
+
+        synchronized (PROPS_LOCK) {
+            for (String token : malformedRecords) {
+                Object remove = PROPS.remove(token);
+                if (SCREENCAST_DEBUG) {
+                    System.err.println("removing malformed record\n" + remove);
+                }
+            }
+
+            store("remove malformed records");
+        }
+    }
+
+    private static void store(String failMsg) {
+        if (!isWritable()) {
+            return;
+        }
+
+        try (OutputStream output = Files.newOutputStream(PROPS_PATH)) {
```

Comment:
Note that this stream is un-buffered (slow)
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/file/Files.html#newOutputStream(java.nio.file.Path,java.nio.file.OpenOption...)
Maybe doesn't matter given how small the file is.

Reply:
Moved to newBufferedWriter/Readed